File: Example/FWFramework/Classes/Module/Test/Service/TestThemeController.m
//
//  TestThemeController.m
//  FWFramework_Example
//
//  Created by wuyong on 2022/8/24.
//  Copyright © 2022 CocoaPods. All rights reserved.
//

#import "TestThemeController.h"
#import "AppSwift.h"
@import FWFramework;

#define TestImage FWIconImage(@"zmdi-var-flower", 24)

@interface TestThemeController () <FWViewController>

@end

@implementation TestThemeController

- (void)didInitialize
{
    [self fw_observeNotification:FWThemeChangedNotification block:^(NSNotification * _Nonnull notification) {
        NSLog(@"主题改变通知：%@", @(FWThemeManager.sharedInstance.style));
    }];
    
    // iOS13以下named方式不支持动态颜色和图像，可手工注册之
    if (@available(iOS 13.0, *)) { } else {
        [UIColor fw_setThemeColor:[UIColor fw_themeLight:[UIColor blackColor] dark:[UIColor whiteColor]] forName:@"themeColor"];
        [UIImage fw_setThemeImage:[UIImage fw_themeLight:[FWModuleBundle imageNamed:@"themeImageLight"] dark:[FWModuleBundle imageNamed:@"themeImageDark"]] forName:@"themeImage"];
    }
}

- (void)setupSubviews
{
    self.fw_extendedLayoutEdge = UIRectEdgeBottom;
    self.view.backgroundColor = [UIColor fw_themeLight:[UIColor whiteColor] dark:[UIColor blackColor]];
    
    UIView *colorView = [[UIView alloc] initWithFrame:CGRectMake(20, 20, 50, 50)];
    colorView.backgroundColor = [UIColor fw_themeLight:[UIColor blackColor] dark:[UIColor whiteColor]];
    [self.view addSubview:colorView];
    
    colorView = [[UIView alloc] initWithFrame:CGRectMake(90, 20, 50, 50)];
    colorView.backgroundColor = [UIColor fw_themeNamed:@"themeColor" bundle:FWModuleBundle.bundle];
    [self.view addSubview:colorView];
    
    UIView *colorView1 = [[UIView alloc] initWithFrame:CGRectMake(160, 20, 50, 50)];
    [UIColor fw_setThemeColor:[UIColor fw_themeLight:[UIColor blackColor] dark:[UIColor whiteColor]] forName:@"dynamicColor"];
    UIColor *dynamicColor = [UIColor fw_themeNamed:@"dynamicColor"];
    colorView1.backgroundColor = [dynamicColor fw_colorForStyle:FWThemeManager.sharedInstance.style];
    FWWeakify(colorView1);
    [colorView1 fw_addThemeListener:^(FWThemeStyle style) {
        FWStrongify(colorView1);
        colorView1.backgroundColor = [dynamicColor fw_colorForStyle:style];
    }];
    [self.view addSubview:colorView1];
    
    UIImageView *imageView1 = [[UIImageView alloc] initWithFrame:CGRectMake(20, 90, 50, 50)];
    UIImage *themeImage = [UIImage fw_themeLight:[FWModuleBundle imageNamed:@"themeImageLight"] dark:[FWModuleBundle imageNamed:@"themeImageDark"]];
    imageView1.image = themeImage.fw_image;
    FWWeakify(imageView1);
    [imageView1 fw_addThemeListener:^(FWThemeStyle style) {
        FWStrongify(imageView1);
        imageView1.image = themeImage.fw_image;
    }];
    [self.view addSubview:imageView1];
    
    UIImageView *imageView = [[UIImageView alloc] initWithFrame:CGRectMake(90, 90, 50, 50)];
    imageView.fw_themeImage = [UIImage fw_themeNamed:@"themeImage" bundle:FWModuleBundle.bundle];
    [self.view addSubview:imageView];
    
    UIImageView *imageView2 = [[UIImageView alloc] initWithFrame:CGRectMake(160, 90, 50, 50)];
    UIImage *reverseImage = [UIImage fw_themeNamed:@"themeImage" bundle:FWModuleBundle.bundle];
    imageView2.image = [reverseImage fw_imageForStyle:FWThemeManager.sharedInstance.style];
    FWWeakify(imageView2);
    [imageView2 fw_addThemeListener:^(FWThemeStyle style) {
        FWStrongify(imageView2);
        imageView2.image = [reverseImage fw_imageForStyle:style];
    }];
    [self.view addSubview:imageView2];
    
    CALayer *layer1 = [CALayer new];
    layer1.frame = CGRectMake(20, 160, 50, 50);
    UIColor *themeColor = [UIColor fw_themeLight:[UIColor blackColor] dark:[UIColor whiteColor]];
    layer1.backgroundColor = [themeColor fw_colorForStyle:FWThemeManager.sharedInstance.style].CGColor;
    layer1.fw_themeContext = self;
    FWWeakify(layer1);
    [layer1 fw_addThemeListener:^(FWThemeStyle style) {
        FWStrongify(layer1);
        layer1.backgroundColor = [themeColor fw_colorForStyle:style].CGColor;
    }];
    [self.view.layer addSublayer:layer1];
    
    CALayer *layer = [CALayer new];
    layer.frame = CGRectMake(90, 160, 50, 50);
    layer.fw_themeContext = self.view;
    layer.fw_themeBackgroundColor = [UIColor fw_themeColor:^UIColor * _Nonnull(FWThemeStyle style) {
        return style == FWThemeStyleDark ? [UIColor whiteColor] : [UIColor blackColor];
    }];
    [self.view.layer addSublayer:layer];
    
    CAGradientLayer *gradientLayer = [CAGradientLayer new];
    gradientLayer.frame = CGRectMake(160, 160, 50, 50);
    gradientLayer.fw_themeContext = self;
    gradientLayer.fw_themeColors = @[[UIColor fw_themeNamed:@"themeColor" bundle:FWModuleBundle.bundle], [UIColor fw_themeNamed:@"themeColor" bundle:FWModuleBundle.bundle]];
    [self.view.layer addSublayer:gradientLayer];
    
    CALayer *layer2 = [CALayer new];
    layer2.frame = CGRectMake(20, 230, 50, 50);
    UIImage *layerImage = [UIImage fw_themeLight:[FWModuleBundle imageNamed:@"themeImageLight"] dark:[FWModuleBundle imageNamed:@"themeImageDark"]];
    layer2.contents = (id)layerImage.fw_image.CGImage;
    layer2.fw_themeContext = self.view;
    FWWeakify(layer2);
    [layer2 fw_addThemeListener:^(FWThemeStyle style) {
        FWStrongify(layer2);
        layer2.contents = (id)layerImage.fw_image.CGImage;
    }];
    [self.view.layer addSublayer:layer2];
    
    layer = [CALayer new];
    layer.frame = CGRectMake(90, 230, 50, 50);
    layer.fw_themeContext = self;
    layer.fw_themeContents = [UIImage fw_themeImage:^UIImage * _Nonnull(FWThemeStyle style) {
        return style == FWThemeStyleDark ? [FWModuleBundle imageNamed:@"themeImageDark"] : [FWModuleBundle imageNamed:@"themeImageLight"];
    }];
    [self.view.layer addSublayer:layer];
    
    layer = [CALayer new];
    layer.frame = CGRectMake(160, 230, 50, 50);
    layer.fw_themeContext = self.view;
    layer.fw_themeContents = [UIImage fw_themeNamed:@"themeImage" bundle:FWModuleBundle.bundle];
    [self.view.layer addSublayer:layer];
    
    imageView = [[UIImageView alloc] initWithFrame:CGRectMake(20, 300, 50, 50)];
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        UIImage.fw_themeImageColor = AppTheme.textColor;
    });
    imageView.fw_themeImage = TestImage.fw_themeImage;
    [self.view addSubview:imageView];
    
    imageView = [[UIImageView alloc] initWithFrame:CGRectMake(90, 300, 50, 50)];
    UIImage *colorImage = [UIImage fw_themeLight:TestImage dark:TestImage];
    imageView.fw_themeImage = [colorImage fw_themeImageWithColor:AppTheme.textColor];
    [self.view addSubview:imageView];
    
    imageView = [[UIImageView alloc] initWithFrame:CGRectMake(160, 300, 50, 50)];
    [UIImage fw_setThemeImage:[UIImage fw_themeLight:TestImage dark:[TestImage fw_imageWithTintColor:[UIColor whiteColor]]] forName:@"dynamicImage"];
    UIImage *dynamicImage = [UIImage fw_themeNamed:@"dynamicImage"];
    imageView.image = [dynamicImage fw_imageForStyle:FWThemeManager.sharedInstance.style];
    FWWeakify(imageView);
    [imageView fw_addThemeListener:^(FWThemeStyle style) {
        FWStrongify(imageView);
        imageView.image = [dynamicImage fw_imageForStyle:style];
    }];
    [self.view addSubview:imageView];
    
    imageView = [[UIImageView alloc] initWithFrame:CGRectMake(20, 370, 50, 50)];
    imageView.fw_themeAsset = [UIImageAsset fw_themeLight:TestImage dark:[TestImage fw_imageWithTintColor:[UIColor whiteColor]]];
    [self.view addSubview:imageView];
    
    imageView = [[UIImageView alloc] initWithFrame:CGRectMake(90, 370, 50, 50)];
    imageView.fw_themeAsset = [UIImageAsset fw_themeAsset:^UIImage * (FWThemeStyle style) {
        return style == FWThemeStyleDark ? [TestImage fw_imageWithTintColor:[UIColor whiteColor]] : TestImage;
    }];
    [self.view addSubview:imageView];
    
    UIImageView *imageView3 = [[UIImageView alloc] initWithFrame:CGRectMake(160, 370, 50, 50)];
    imageView3.fw_themeImage = [TestImage fw_themeImageWithColor:[UIColor redColor]];
    [self.view addSubview:imageView3];
    
    UILabel *colorLabel = [UILabel new];
    colorLabel.frame = CGRectMake(0, 440, FWScreenWidth, 25);
    colorLabel.textAlignment = NSTextAlignmentCenter;
    colorLabel.font = FWFontSize(16).fw_boldFont;
    colorLabel.textColor = AppTheme.textColor;
    UIColor *lightColor = [AppTheme.textColor fw_colorForStyle:FWThemeStyleLight];
    UIColor *darkColor = [AppTheme.textColor fw_colorForStyle:FWThemeStyleDark];
    colorLabel.text = [NSString stringWithFormat:@"Light: %@ Dark: %@", lightColor.fw_hexString, darkColor.fw_hexString];
    [self.view addSubview:colorLabel];
    
    UILabel *themeLabel = [UILabel new];
    themeLabel.frame = CGRectMake(0, 475, FWScreenWidth, 25);
    themeLabel.textAlignment = NSTextAlignmentCenter;
    themeLabel.attributedText = [NSAttributedString fw_attributedString:@"我是AttributedString" withFont:FWFontSize(16).fw_boldFont textColor:[UIColor fw_themeLight:[UIColor blackColor] dark:[UIColor whiteColor]]];
    [self.view addSubview:themeLabel];
    
    UIButton *themeButton = [UIButton new];
    themeButton.frame = CGRectMake(0, 510, FWScreenWidth, 50);
    themeButton.titleLabel.font = FWFontRegular(16);
    [themeButton setTitleColor:[UIColor fw_themeLight:[UIColor blackColor] dark:[UIColor whiteColor]] forState:UIControlStateNormal];
    
    UIImage *buttonImage = [UIImage fw_themeLight:(FWThemeManager.sharedInstance.style == FWThemeStyleLight ? nil : [FWModuleBundle imageNamed:@"themeImageLight"]) dark:(FWThemeManager.sharedInstance.style == FWThemeStyleDark ? nil : [FWModuleBundle imageNamed:@"themeImageDark"])];
    FWThemeObject<NSAttributedString *> *themeString = [NSAttributedString fw_themeObjectWithHtmlString:@"我是<span style='color:red;'>红色</span>AttributedString" defaultAttributes:@{
        NSFontAttributeName: FWFontBold(16),
        NSForegroundColorAttributeName: [UIColor fw_themeLight:[UIColor blackColor] dark:[UIColor whiteColor]],
    }];
    
    [themeButton setImage:buttonImage.fw_image forState:UIControlStateNormal];
    [themeButton setAttributedTitle:themeString.object forState:UIControlStateNormal];
    [themeLabel fw_addThemeListener:^(FWThemeStyle style) {
        [themeButton setImage:buttonImage.fw_image forState:UIControlStateNormal];
        [themeButton setAttributedTitle:themeString.object forState:UIControlStateNormal];
    }];
    [self.view addSubview:themeButton];
}

- (void)setupNavbar
{
    FWThemeMode mode = FWThemeManager.sharedInstance.mode;
    NSString *title = mode == FWThemeModeSystem ? @"系统" : (mode == FWThemeModeDark ? @"深色" : @"浅色");
    FWWeakifySelf();
    [self fw_setRightBarItem:title block:^(id  _Nonnull sender) {
        FWStrongifySelf();
        
        NSMutableArray *actions = [NSMutableArray arrayWithArray:@[@"系统", @"浅色", @"深色"]];
        [self fw_showSheetWithTitle:nil message:nil cancel:@"取消" actions:actions actionBlock:^(NSInteger index) {
            FWStrongifySelf();
            
            FWThemeManager.sharedInstance.mode = index;
            [self setupNavbar];
            id appDelegate = UIApplication.sharedApplication.delegate;
            if ([appDelegate isKindOfClass:[FWAppDelegate class]]) {
                [(FWAppDelegate *)appDelegate setupController];
            }
        }];
    }];
}

@end

File: Example/FWFramework/Classes/Module/Test/Service/TestThemeController.swift
//
//  TestThemeController.swift
//  FWFramework_Example
//
//  Created by wuyong on 2022/9/15.
//  Copyright © 2022 CocoaPods. All rights reserved.
//

import FWFramework

class TestThemeController: UIViewController, ViewControllerProtocol {
    static let testImage = APP.iconImage("zmdi-var-flower", 24)

    func didInitialize() {
        app.observeNotification(.ThemeChanged) { _ in
            APP.debug("主题改变通知：\(ThemeManager.shared.style.rawValue)")
        }
    }

    func setupSubviews() {
        app.extendedLayoutEdge = .bottom
        view.backgroundColor = UIColor.app.themeLight(.white, dark: .black)

        var colorView = UIView(frame: CGRect(x: 20, y: 20, width: 50, height: 50))
        colorView.backgroundColor = UIColor.app.themeLight(.black, dark: .white)
        view.addSubview(colorView)

        colorView = UIView(frame: CGRect(x: 90, y: 20, width: 50, height: 50))
        colorView.backgroundColor = UIColor.app.themeNamed("themeColor", bundle: ModuleBundle.bundle())
        view.addSubview(colorView)

        let colorView1 = UIView(frame: CGRect(x: 160, y: 20, width: 50, height: 50))
        UIColor.app.setThemeColor(UIColor.app.themeLight(.black, dark: .white), forName: "dynamicColor")
        let dynamicColor = UIColor.app.themeNamed("dynamicColor")
        colorView1.backgroundColor = dynamicColor.app.color(forStyle: ThemeManager.shared.style)
        colorView1.app.addThemeListener { style in
            colorView1.backgroundColor = dynamicColor.app.color(forStyle: style)
        }
        view.addSubview(colorView1)

        let imageView1 = UIImageView(frame: CGRect(x: 20, y: 90, width: 50, height: 50))
        let themeImage = UIImage.app.themeLight(ModuleBundle.imageNamed("themeImageLight"), dark: ModuleBundle.imageNamed("themeImageDark"))
        imageView1.image = themeImage.app.image
        imageView1.app.addThemeListener { _ in
            imageView1.image = themeImage.app.image
        }
        view.addSubview(imageView1)

        var imageView = UIImageView(frame: CGRect(x: 90, y: 90, width: 50, height: 50))
        imageView.app.themeImage = UIImage.app.themeNamed("themeImage", bundle: ModuleBundle.bundle())
        view.addSubview(imageView)

        let imageView2 = UIImageView(frame: CGRect(x: 160, y: 90, width: 50, height: 50))
        let reverseImage = UIImage.app.themeNamed("themeImage", bundle: ModuleBundle.bundle())
        imageView2.image = reverseImage.app.image(forStyle: ThemeManager.shared.style)
        imageView2.app.addThemeListener { style in
            imageView2.image = reverseImage.app.image(forStyle: style)
        }
        view.addSubview(imageView2)

        let layer1 = CALayer()
        layer1.frame = CGRect(x: 20, y: 160, width: 50, height: 50)
        let themeColor = UIColor.app.themeLight(.black, dark: .white)
        layer1.backgroundColor = themeColor.app.color(forStyle: ThemeManager.shared.style).cgColor
        layer1.app.themeContext = self
        layer1.app.addThemeListener { style in
            layer1.backgroundColor = themeColor.app.color(forStyle: style).cgColor
        }
        view.layer.addSublayer(layer1)

        var layer = CALayer()
        layer.frame = CGRect(x: 90, y: 160, width: 50, height: 50)
        layer.app.themeContext = view
        layer.app.themeBackgroundColor = UIColor.app.themeColor { style in
            style == .dark ? .white : .black
        }
        view.layer.addSublayer(layer)

        let gradientLayer = CAGradientLayer()
        gradientLayer.frame = CGRect(x: 160, y: 160, width: 50, height: 50)
        gradientLayer.app.themeContext = self
        gradientLayer.app.themeColors = [UIColor.app.themeNamed("themeColor", bundle: ModuleBundle.bundle()), UIColor.app.themeNamed("themeColor", bundle: ModuleBundle.bundle())]
        view.layer.addSublayer(gradientLayer)

        let layer2 = CALayer()
        layer2.frame = CGRect(x: 20, y: 230, width: 50, height: 50)
        let layerImage = UIImage.app.themeLight(ModuleBundle.imageNamed("themeImageLight"), dark: ModuleBundle.imageNamed("themeImageDark"))
        layer2.contents = layerImage.app.image?.cgImage
        layer2.app.themeContext = view
        layer2.app.addThemeListener { _ in
            layer2.contents = layerImage.app.image?.cgImage
        }
        view.layer.addSublayer(layer2)

        layer = CALayer()
        layer.frame = CGRect(x: 90, y: 230, width: 50, height: 50)
        layer.app.themeContext = self
        layer.app.themeContents = UIImage.app.themeImage { style in
            style == .dark ? ModuleBundle.imageNamed("themeImageDark") : ModuleBundle.imageNamed("themeImageLight")
        }
        view.layer.addSublayer(layer)

        layer = CALayer()
        layer.frame = CGRect(x: 160, y: 230, width: 50, height: 50)
        layer.app.themeContext = view
        layer.app.themeContents = UIImage.app.themeNamed("themeImage", bundle: ModuleBundle.bundle())
        view.layer.addSublayer(layer)

        imageView = UIImageView(frame: CGRect(x: 20, y: 300, width: 50, height: 50))
        UIImage.app.themeImageColorConfiguration = { AppTheme.textColor }
        imageView.app.themeImage = Self.testImage?.app.themeImage
        view.addSubview(imageView)

        imageView = UIImageView(frame: CGRect(x: 90, y: 300, width: 50, height: 50))
        let colorImage = UIImage.app.themeLight(Self.testImage, dark: Self.testImage)
        imageView.app.themeImage = colorImage.app.themeImage(color: AppTheme.textColor)
        view.addSubview(imageView)

        imageView = UIImageView(frame: CGRect(x: 160, y: 300, width: 50, height: 50))
        UIImage.app.setThemeImage(UIImage.app.themeLight(Self.testImage, dark: Self.testImage?.app.image(tintColor: .white)), forName: "dynamicImage")
        let dynamicImage = UIImage.app.themeNamed("dynamicImage")
        imageView.image = dynamicImage.app.image(forStyle: ThemeManager.shared.style)
        imageView.app.addThemeListener { style in
            imageView.image = dynamicImage.app.image(forStyle: style)
        }
        view.addSubview(imageView)

        imageView = UIImageView(frame: CGRect(x: 20, y: 370, width: 50, height: 50))
        imageView.app.themeAsset = UIImageAsset.app.themeLight(Self.testImage, dark: Self.testImage?.app.image(tintColor: .white))
        view.addSubview(imageView)

        imageView = UIImageView(frame: CGRect(x: 90, y: 370, width: 50, height: 50))
        imageView.app.themeAsset = UIImageAsset.app.themeAsset { style in
            style == .dark ? Self.testImage?.app.image(tintColor: .white) : Self.testImage
        }
        view.addSubview(imageView)

        let imageView3 = UIImageView(frame: CGRect(x: 160, y: 370, width: 50, height: 50))
        imageView3.app.themeImage = Self.testImage?.app.themeImage(color: .red)
        view.addSubview(imageView3)

        let colorLabel = UILabel()
        colorLabel.frame = CGRect(x: 0, y: 440, width: APP.screenWidth, height: 25)
        colorLabel.textAlignment = .center
        colorLabel.font = APP.font(16).app.boldFont
        colorLabel.textColor = AppTheme.textColor
        let lightColor = AppTheme.textColor.app.color(forStyle: .light)
        let darkColor = AppTheme.textColor.app.color(forStyle: .dark)
        colorLabel.text = "Light: \(lightColor.app.hexString) Dark: \(darkColor.app.hexString)"
        view.addSubview(colorLabel)

        let themeLabel = UILabel()
        themeLabel.frame = CGRect(x: 0, y: 475, width: APP.screenWidth, height: 25)
        themeLabel.textAlignment = .center
        themeLabel.attributedText = NSAttributedString.app.attributedString("我是AttributedString", font: APP.font(16).app.boldFont, textColor: UIColor.app.themeLight(.black, dark: .white))
        view.addSubview(themeLabel)

        let themeButton = UIButton()
        themeButton.frame = CGRect(x: 0, y: 510, width: APP.screenWidth, height: 50)
        themeButton.titleLabel?.font = APP.font(16)
        themeButton.setTitleColor(UIColor.app.themeLight(.black, dark: .white), for: .normal)

        let buttonImage = UIImage.app.themeLight(ThemeManager.shared.style == .light ? nil : ModuleBundle.imageNamed("themeImageLight"), dark: ThemeManager.shared.style == .dark ? nil : ModuleBundle.imageNamed("themeImageDark"))
        let themeString = NSAttributedString.app.themeObject(htmlString: "我是<span style='color:red;'>红色</span>AttributedString", defaultAttributes: [
            .font: APP.font(16, .bold),
            .foregroundColor: UIColor.app.themeLight(.black, dark: .white)
        ])

        themeButton.setImage(buttonImage.app.image, for: .normal)
        themeButton.setAttributedTitle(themeString.object, for: .normal)
        themeLabel.app.addThemeListener { _ in
            themeButton.setImage(buttonImage.app.image, for: .normal)
            themeButton.setAttributedTitle(themeString.object, for: .normal)
        }
        view.addSubview(themeButton)
    }
}
